### PR TITLE
[ch179] Add files to connect to Docker

### PR DESCRIPTION
**Pull request by Raksani**
https://app.clubhouse.io/cobda/story/179/automated-build-docker-image

## Why?

To build Docker image we need Dockerfile to run the command so that every update on branch `development` will activate auto-build Docker image on Docker Hub. And processes are available to real-time monitor on Docker Hub repository.
Docker Hub: https://hub.docker.com/repository/docker/cobda/cobda

## Changes

- Our website is ready to use as **Docker Image** on **Docker Hub**
- Continuous Deployment is working successfully

## Screenshots
Config build rules
Screenshot 2564-02-10 at 20 46 11

After click `save and build`, automated builds log can be monitored via Docker Hub repository 
Screenshot 2564-02-10 at 20 46 28
Screenshot 2564-02-10 at 20 46 50
Screenshot 2564-02-10 at 20 47 58

Build Success
Screenshot 2564-02-15 at 00 02 06
Show on specific port (:8000)
Screenshot 2564-02-10 at 20 14 45

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
+FROM node:current-alpine AS base
+WORKDIR /base
+COPY package*.json ./
+RUN npm install
```
Comment (LilBank) [inline Dockerfile]: I was wondering if we can use `yarn` to speed up the overall process. This [link](https://medium.com/hackernoon/using-yarn-with-docker-c116ad289d56) provide instructions while this [link](https://medium.com/@iamnayr/a-multi-part-analysis-of-node-docker-image-sizes-using-yarn-vs-traditional-npm-2c20f034c08f) shows the comparison. Please kindly have a look and if possible let's use `yarn`.

File: .dockerignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+.next
```
Comment (LilBank) [inline .dockerignore]: Are we suppose to add `node_modules` too? How about `.github` and `.eslintcache`? Try turning on hidden files in your editor and see which files should be ignored.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,23 @@
+FROM node:current-alpine AS base
```
Comment (LilBank) [inline Dockerfile]: Would be nice to name this stages 🥰 
See [here](https://docs.docker.com/develop/develop-images/multistage-build/#name-your-build-stages)

**Review by LilBank — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (Raksani) [reply Dockerfile]: Very interesting, I decide to use `npm` as a default so that our `dockerfile` is surely worked. Anyway, I've to modify this  a bit to support `yarn`.

File: Dockerfile (hunk 3 above)
Comment (Raksani) [reply Dockerfile]: From the [Docker document](https://docs.docker.com/develop/develop-images/multistage-build/#name-your-build-stages). "By default, the stages are not named. However, you can name your stages, by adding an AS <NAME>" So I named it as `BASE`. Anyway, it's really optional.

File: .dockerignore [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+.node_repl_history
+
+# Output of 'npm pack'
+*.tgz
+
+# Yarn Integrity file
+.yarn-integrity
+
+# Git
+.git
```
Comment (LilBank) [inline .dockerignore]: Please add a newline.

File: .dockerignore [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+build/Release
+
+# Typescript v1 declaration files
+typings/
+
+# Optional npm cache directory
+.npm
+
+# Optional eslint cache
+.eslintcache
```
Comment (LilBank) [inline .dockerignore]: Quick reminder, I think this should already be ignored by `.gitignore` so no need to put this here. Please also check for other duplicates.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+# Production image, copy all the files and run next
+FROM node:current-alpine AS runner
+ENV NODE_ENV=production
+WORKDIR /app
+
+# Copy from build image
+COPY --from=builder /build/package*.json ./
+COPY --from=builder /build/node_modules ./node_modules
+COPY --from=builder /build/.next ./.next
+COPY --from=builder /build/public ./public
```
Comment (LilBank) [inline Dockerfile]: This should leave us 4 layers of `COPY`. According to [here](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers), we should reduce number of layers and chain `COPY` instead by [this](https://stackoverflow.com/questions/30256386/how-to-copy-multiple-files-in-one-layer-using-a-dockerfile).

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+# Expose the listening port
+EXPOSE 3000
+# Run yarn start script when container starts
+CMD ["yarn", "start"]
```
Comment (LilBank) [inline Dockerfile]: Please add a newline.

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+# Set working directory
+WORKDIR /base
+
+# Copy package.json and package-lock.json before other files
+# Utilise Docker cache to save re-installing dependencies if unchanged
+COPY package*.json yarn.lock ./ 
+
+# Install dependencies
+RUN yarn install --frozen-lockfile
```
Comment (LilBank) [inline Dockerfile:13]: Quick question, should we also pass `--production` too?

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+# Copy all files
```
Comment (LilBank) [inline Dockerfile]: How about this?
```suggestion
# Copy installed dependencies to current working directory
```

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,39 @@
+# Base on offical Node.js Alpine image
+FROM node:current-alpine AS base
+
+# Set working directory
+WORKDIR /base
+
+# Copy package.json and package-lock.json before other files
```
Comment (LilBank) [inline Dockerfile]:
```suggestion
# Copy package.json and yarn-lock.json to [SOME_PATH] for installation
```

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+# Utilise Docker cache to save re-installing dependencies if unchanged
```
Comment (LilBank) [inline Dockerfile]:
```suggestion
# This is to utilise Docker cache to save re-installing dependencies if unchanged
```

File: Dockerfile (hunk 8 above)
Comment (LilBank) [reply Dockerfile:13]: Quick question, since this is the part where we install dependencies, should we also add `ENV NODE_ENV=production` in this stage?

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+COPY package*.json yarn.lock ./ 
+
+# Install dependencies
+RUN yarn install --frozen-lockfile
+
+# Copy all files
+COPY . .
+
+# Build app
+FROM base AS builder
```
Comment (LilBank) [inline Dockerfile:19]: Since above we add the comment to explain each step, would be nice to have it here too 🥰

File: Dockerfile [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+ENV NODE_ENV=production
+WORKDIR /build
+COPY . .
+COPY --from=base /base/node_modules ./node_modules
```
Comment (LilBank) [inline Dockerfile]: If my understandings are correct, the above command is for copying everything in the `base` to `build`. Then why are you copying the `node_modules` again? 🤔

File: Dockerfile [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+# Copy from build image
+COPY --from=builder /build/package*.json ./
+COPY --from=builder /build/node_modules ./node_modules
+COPY --from=builder /build/.next ./.next
+COPY --from=builder /build/public ./public
+
+# Expose the listening port
+EXPOSE 3000
+# Run yarn start script when container starts
```
Comment (LilBank) [inline Dockerfile]:
```suggestion
EXPOSE 3000

# Run yarn start script when container starts
```

File: Dockerfile [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+COPY . .
+COPY --from=base /base/node_modules ./node_modules
+RUN yarn build
+
+# Production image, copy all the files and run next
+FROM node:current-alpine AS runner
+ENV NODE_ENV=production
+WORKDIR /app
+
+# Copy from build image
```
Comment (LilBank) [inline Dockerfile]: Can you please make this sentence more meaningful?
```suggestion
# Copy built files to ...
```

File: Dockerfile [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+# Utilise Docker cache to save re-installing dependencies if unchanged
+COPY package*.json yarn.lock ./ 
+
+# Install dependencies
+RUN yarn install --frozen-lockfile
+
+# Copy all files
+COPY . .
+
+# Build app
```
Comment (LilBank) [inline Dockerfile]:
```suggestion
# Build image, for building the app.
```

File: Dockerfile [hunk 17]
```diff
@@ -0,0 +1,39 @@
+# Base on offical Node.js Alpine image
```
Comment (LilBank) [inline Dockerfile]:
```suggestion
# Base image, for installing the dependencies.
```

File: Dockerfile [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+# Build app
+FROM base AS builder
+ENV NODE_ENV=production
+WORKDIR /build
+COPY . .
+COPY --from=base /base/node_modules ./node_modules
+RUN yarn build
+
+# Production image, copy all the files and run next
```
Comment (LilBank) [inline Dockerfile]:
```suggestion
# Production image, [DESCRIPTION]
```

**Review by LilBank — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 11 above)
Comment (Raksani) [reply Dockerfile]: Just `Utilize` is fine. Like `Copy`

Comment (LilBank) [reply Dockerfile]: Utilize or Utilise?

File: Dockerfile (hunk 6 above)
Comment (Raksani) [reply Dockerfile]: `COPY` them individually, rather than all at once. This ensures that each step’s build successfully. It’s easy to maintain and monitor. I prefer to keep it like this. You can see more examples in file changed of this [PR](https://github.com/vercel/next.js/pull/17794)

Comment (LilBank) [reply Dockerfile]: According to your reference, author only have 2 copies while we have 4. This is only a `COPY` command which copies the built file to the `/app` which I still don't think it should be done separately. Not like `RUN` command which some should be separated for monitoring though. What do you think? Please share some thoughts @kimvcp .

BTW, can't we just copy all the files in `builder` and just paste in `WORKDIR`?

Comment (yordis) [reply Dockerfile]: @LilBank is a balance between running the entire stage even though you only need a section of it, vs having multi-stage where it may be the case you copy things multiple times, but it is optimized for caching the layers.

So technically, depends on which layer of invalidation you are producing.

I would say, test it if you think it is slower, this is the best I came up with https://github.com/vercel/next.js/discussions/16995#discussioncomment-132339

Comment (LilBank) [reply Dockerfile]: Really appreciate your comment 🙇🏻‍♂️. However, I think my [comment](https://github.com/Cobda/cobda-web/pull/17/files#discussion_r575370998) is unclear so let me explain it explicitly.

My first thought was neither of each approach are slower than each other (multi `COPY`s/single `COPY`) and when I tested them the results were as expected. I have this thought because I was trying to follow **Dockerfile best practices** documentation([here](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/)). The document mention that we should **minimize** the number of layers in the image([see here](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers)) since `ADD` and `COPY` command creates a layer. Furthermore, both approach caches will be invalid([here, 3rd bullet point](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache)) since both depends on the **built files** from the `builder` image (file changed => new commit => new build = (`cache COPY =! new COPY`) ).

So this is my thought 😅

In conclusion, both approach gives the same performance but only the generated layers are the diff. What do you think about this? Should we separate it into thin layers of one thick layer? 🤔 @yordis

PS. I'm quite new to Docker so if I miss understood anything feel free to correct me 🥰

File: Dockerfile (hunk 11 above)
Comment (Raksani) [reply Dockerfile]: Utilise

File: Dockerfile (hunk 15 above)
Comment (Raksani) [reply Dockerfile]: I put more detail like this, what do you think? 🤔 
```suggestion
# Copy files from build stage to production stage
```

File: Dockerfile (hunk 16 above)
Comment (Raksani) [reply Dockerfile]: I'll make it clearer, what do you think? 🤔
```suggestion
# Build stage
```

File: Dockerfile (hunk 18 above)
Comment (Raksani) [reply Dockerfile]: I'd make it's synchronized to what I've mentioned in earlier comments. What do you think? 🤔 
```suggestion
# Production stage
```

File: Dockerfile (hunk 8 above)
Comment (Raksani) [reply Dockerfile:12]: Yes, we should do so in every stage. I will put it in.

File: Dockerfile (hunk 10 above)
Comment (Raksani) [reply Dockerfile]: I'd like to explain that "we have to copy package.json and package-lock.json before other files and utilise Docker cache to save re-installing dependencies if unchanged"

File: Dockerfile (hunk 12 above)
Comment (Raksani) [reply Dockerfile:19]: Of course, I will do the same as https://github.com/Cobda/cobda-web/pull/17#discussion_r575689629

File: Dockerfile (hunk 15 above)
Comment (LilBank) [reply Dockerfile]: LGTM

File: Dockerfile (hunk 18 above)
Comment (LilBank) [reply Dockerfile]: SGTM

File: Dockerfile (hunk 16 above)
Comment (LilBank) [reply Dockerfile]: SGTM

File: Dockerfile (hunk 10 above)
Comment (LilBank) [reply Dockerfile]: How about, **we only copy package.json and yarn-lock.json for utilizing Docker cache ...**

File: Dockerfile (hunk 13 above)
Comment (Raksani) [reply Dockerfile]: I'll modify it to be `COPY` all files from `build stage` instead.

File: Dockerfile (hunk 17 above)
Comment (Raksani) [reply Dockerfile]: So, I decided to detail every stage. I'd be like this. What do you think? 🤔💭 
```suggestion
# Installation stage
# Use Node.js Alpine base image
```

File: Dockerfile (hunk 8 above)
Comment (Raksani) [reply Dockerfile:13]: About `NODE_ENV=production`, because this (build) step only relates to a production build, so we're explicitly setting the `NODE_ENV` to production.

File: Dockerfile [hunk 19]
```diff
@@ -0,0 +1,37 @@
+# Installation stage
+# Offical Node.js Alpine base image
+FROM node:lts-alpine AS base
+
+# Set working directory
+WORKDIR /base
+
+# Copy package.json and package-lock.json for utilising Docker cache
```
Comment (LilBank) [inline Dockerfile:8]:
```suggestion
# Copy package.json and yarn-lock.json for utilizing Docker cache 
```

File: Dockerfile [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+# Copy installed dependencies to current working directory
+COPY . .
+
+# Build stage
+FROM base AS builder
+ENV NODE_ENV=production
+WORKDIR /build
+COPY --from=base /base ./
+RUN yarn add --dev typescript @types/react @types/node && yarn build
```
Comment (LilBank) [inline Dockerfile]: It seems like you've added the `devDependencies` here because of `--production` in `yarn install` command. Should we remove the `--production` so this `RUN` command only has `yarn build`?

File: Dockerfile [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+
+# Install dependencies only needed
+RUN yarn install --production --frozen-lockfile
+
+# Copy installed dependencies to current working directory
+COPY . .
+
+# Build stage
+FROM base AS builder
+ENV NODE_ENV=production
```
Comment (LilBank) [inline Dockerfile]: FYI, I've tested that currently the `ENV` isn't needed. We can remove for now.

File: Dockerfile [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+COPY --from=base /base ./
+RUN yarn add --dev typescript @types/react @types/node && yarn build
+
+# Production stage
+FROM node:current-alpine AS runner
+ENV NODE_ENV=production
+WORKDIR /app
+
+# Copy files from build stage to production stage
+COPY --from=builder /build/package*.json /build/node_modules /build/.next /build/public ./
```
Comment (LilBank) [inline Dockerfile:39]: Quick question, couldn't we just do this? Since we're copying all the files from `/build`.
```suggestion
COPY --from=builder /build .
```

File: Dockerfile [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+RUN yarn install --production --frozen-lockfile
+
+# Copy installed dependencies to current working directory
+COPY . .
+
+# Build stage
+FROM base AS builder
+ENV NODE_ENV=production
+WORKDIR /build
+COPY --from=base /base ./
```
Comment (LilBank) [inline Dockerfile]: A slash isn't needed since we specify only one location 😅
```suggestion
COPY --from=base /base .
```

**Review by LilBank — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 20 above)
Comment (Raksani) [reply Dockerfile]: First, I used only `yarn build` command but it cause an error and failed to build image. If you would like to see the detail, you can check in our Dockerhub repository.

File: Dockerfile (hunk 21 above)
Comment (Raksani) [reply Dockerfile:20]: I've set it in case we create `docker-compose` in further development.

Comment (LilBank) [reply Dockerfile]: If it's for future implementation then lets remove ir for now since it's unused. 😇

File: Dockerfile (hunk 23 above)
Comment (Raksani) [reply Dockerfile:22]: I intend to copy to `./` like I copied in `installation stage`

File: Dockerfile (hunk 22 above)
Comment (Raksani) [reply Dockerfile:39]: According to`Docker Document` you mentioned me earlier and my mentor’s suggestion, we can do like you suggested. But if concern about clean code and readability (so we know exactly which is used in production) we should keep it like this. I agree with both ways. So, what do you think?

File: Dockerfile (hunk 23 above)
Comment (Raksani) [reply Dockerfile:22]: Actually, I intend to copy it to `./` like in `installation stage`

File: .dockerignore [hunk 24]
```diff
@@ -0,0 +1,12 @@
+Dockerfile
```
Comment (LilBank) [inline .dockerignore:2]: Another comment for Docker please.

File: Dockerfile [hunk 25]
```diff
@@ -0,0 +1,43 @@
+# Installation stage
+# Offical Node.js Alpine base image
+FROM node:lts-alpine AS base
+
+# Set working directory
+WORKDIR /base
+
+# Copy package.json and package-lock.json for utilising Docker cache
```
Comment (LilBank) [inline Dockerfile]:
```suggestion
# Copy package.json and yarn-lock.json for utilising Docker cache 
```

File: Dockerfile [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+FROM node:lts-alpine AS base
+
+# Set working directory
+WORKDIR /base
+
+# Copy package.json and package-lock.json for utilising Docker cache 
+# to save re-installing dependencies if unchanged
+COPY package*.json yarn.lock ./ 
+
+# Install dependencies only needed
```
Comment (LilBank) [inline Dockerfile]:
```suggestion
# Install dependencies
```

File: Dockerfile [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+RUN yarn install --frozen-lockfile
+
+# Copy installed dependencies to current working directory
```
Comment (LilBank) [inline Dockerfile]:
```suggestion
# Copy installed dependencies to the current working directory
```

File: Dockerfile [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+COPY . .
+
+# Build stage
+FROM base AS builder
+
+# TO DO in further development
+# ENV NODE_ENV=production
+
+WORKDIR /build
+COPY --from=base /base .
```
Comment (LilBank) [inline Dockerfile:27]: Since you add a comment about this on `base` and `production` stage. We should add a comment for this too.

**Review by LilBank — CHANGES_REQUESTED** [review]

File: .dockerignore (hunk 24 above)
Comment (Raksani) [reply .dockerignore:2]: Sure.

File: Dockerfile (hunk 25 above)
Comment (Raksani) [reply Dockerfile:8]: So, do you mean this? 
```suggestion
# Copy package.json, package-lock.json, and yarn.lock for utilising Docker cache 
```

File: .dockerignore (hunk 24 above)
Comment (Raksani) [reply .dockerignore:1]: Sure.

File: Dockerfile (hunk 28 above)
Comment (Raksani) [reply Dockerfile:27]: If you mean the comment about `ENV NODE_ENV`, in this file, I add in `build` and `production` stage as in https://github.com/Cobda/cobda-web/pull/17#discussion_r575830993

File: Dockerfile (hunk 27 above)
Comment (Raksani) [reply Dockerfile:15]: So, do you mean this? 
```suggestion
# Copy package.json, package-lock.json, and yarn.lock for utilising Docker cache 
```

File: Dockerfile (hunk 25 above)
Comment (Raksani) [reply Dockerfile:8]: So, do you mean this? 
```suggestion
# Copy package.json, package-lock.json, and yarn.lock for utilising Docker cache 
```

**Comment by Raksani** [comment]
According to https://github.com/Cobda/cobda-web/pull/17#discussion_r575326162, we raise these questions: 
- **Do multiple lines and a single line of the `COPY` command matter for caching in Docker?** 
- **Why don't we put every file from `build stage` to `production stage` in just one simple line?** 
In conclusion, we have 3 criteria to concern:

1. **Image layers' size**: from the experiment, the result from Dockerhub apparently says **no different** between multiple lines and a single line. 

Result: Single-line command 
Screenshot_2564-02-15_at_17 19 46

Result: Multiple-line command
Screenshot_2564-02-15_at_17 37 28

2. **Cache**: Docker looks for an existing image in its cache that it can reuse, rather than creating a new (duplicate) image. Technically, it builds from top to bottom. This means when `package.json` is changed, it will rebuild the first line `COPY --from=builder /build/package*.json ./` including the 3 following commands also which caused time consumption.
3. **Readability**: multiple-line command provide more detail than `COPY from--builder /build/ ./` in single line. So, as the cache is concerned, we decide to specify each package but still in one line.

File: Dockerfile (hunk 28 above)
Comment (LilBank) [reply Dockerfile:27]: Nope, I mean to add this `# Copy files from base stage to builder stage`

Comment (Raksani) [reply Dockerfile:27]: Okay, I'll add for more clarification.

File: Dockerfile [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+
+# Production stage
+FROM node:current-alpine AS runner
+
+# TO DO in further development
+# ENV NODE_ENV=production
+
+WORKDIR /app
+
+# Copy files from build stage to production stage
```
Comment (LilBank) [inline Dockerfile:38]:
```suggestion
# Copy files from builder stage to production stage
```

**Review by LilBank — APPROVED** [review]

**Review by Kittinske15 — APPROVED** [review]

**Review by kimvcp — APPROVED** [review]
